### PR TITLE
Rearrange "SDM Concepts" page

### PR DESCRIPTION
**Pull request by gjtorikian**
One of the first things that threw me off when reading through the Atomist docs was this "SDM Concepts" page. Given the title, I had thought that it would focus on the core components of an SDM. It does, but it also jumps around between code samples and API signatures, which made following the "What _is_ an SDM?" narrative harder to follow.

In this PR, I rearranged a lot of the content. I wanted to keep the SDM Concepts page strictly text--no code samples. In the need for calling out capabilities, like with GraphQL, I moved the content to an existing page. In the need for preserving code samples, I moved those to their own stand-alone article snippets. 

In short, I personally think this page works best by describing the components of what makes an SDM, how it runs, what it can do, etc., and then link out to other docs for readers to get more familiarity.

### CODE REVIEW

File: docs/developer/transform.md [hunk 1]
```diff
@@ -15,6 +15,16 @@ This page shows how to:
 
 After that, you might want to make your code transform into an [Autofix](autofix.md).
 
+There's also an important capability called "dry run transform":
+Performing a transform on a branch, and then either raising either a PR or
+an issue, depending on build success or failure. This allows us to
+safely apply transforms across many repositories. There's a simple wrapper
+function to enable this.
```
Comment (gjtorikian) [inline docs/developer/transform.md]: This is a sentence needing more information, but [the original text](https://github.com/atomist/docs/blob/bdada7b458e833b44a5458ec1eafdf4e663329dd/docs/developer/sdm-concepts.md#dry-run-transforms) also abruptly ended in this way. What is the simple wrapper function we can link to here?

File: docs/glossary.md [hunk 2]
```diff
@@ -6,14 +6,6 @@
  the part of the Atomist platform that is operated by Atomist the company:
  the event hub, GraphQL endpoint, and web interface. [more info](developer/architecture.md)
 
-#### PushRule
-
-a specification for goals to set on a given push. [more info](developer/set-goals.md#pushrule)
```
Comment (gjtorikian) [inline docs/glossary.md]: Not sure why these were here, but I alphabetically moved these closer to the `p`s. Maybe because capital letters are alphabetized ahead of lowercase ones?

File: docs/developer/set-goals.md [hunk 3]
```diff
@@ -79,11 +79,47 @@ The DSL includes support for logical operations. For example, this will build al
 whenPushSatisfies(IsMaven, not(IsKotlin)).setGoals(BuildGoals),
 ```
 
-## PushRule
-
 Each argument to [`sdm.withPushRules`](https://atomist.github.io/sdm/interfaces/_lib_api_machine_softwaredeliverymachine_.softwaredeliverymachine.html#withpushrules)
 is a PushRule, contributing goals on a commit if a condition is met. That condition is a PushTest.
 
+### Advanced Push Rules
```
Comment (gjtorikian) [inline docs/developer/set-goals.md]: I think linking to a section specifically on push rules would make sense here, but we really only talk about [push tests](https://docs.atomist.com/developer/push-test/). Should that article be made more generic to talk about how push tests can be a component of a push rule?

File: docs/glossary.md (hunk 2 above)
Comment (jessitron) [reply docs/glossary.md]: oh dear. This is alphabetized by the SDM, so I need to change that

File: docs/developer/graphql.md [hunk 4]
```diff
@@ -5,6 +5,18 @@ Besides being a great query language, GraphQL provides great tool support based
...
 
+The Atomist automation API provides you access to the events and data
+from your development platforms using [GraphQL][graphql], a
+widely-used query language and runtime for APIs.
+
+You can use GraphQL with the Atomist automation API for:
+
+1.  Queries that fetch data directly
+2.  Subscriptions to register the types of events you want to receive
+3.  Mutations to change data and make connections
```
Comment (jessitron) [inline docs/developer/graphql.md]: is this true? We have very limited mutations available AFAIK. I wouldn't imply that there is a general ability to do this (unless @kipz says otherwise)

File: docs/developer/graphql.md [hunk 5]
```diff
@@ -236,3 +248,57 @@ import * as graphql from "./typings/types";
...
+
+```typescript
+@EventHandler("Set status on build complete",
+    GraphQL.subscriptionFromFile("graphql/subscription/OnBuildComplete.graphql"))
+export class SetStatusOnBuildComplete implements HandleEvent<OnBuildComplete.Subscription> {
+
+    public async handle(event: EventFired<OnBuildComplete.Subscription>,
+    	ctx: HandlerContext,
+    	params: this): Promise<HandlerResult> {
+```
```
Comment (jessitron) [inline docs/developer/graphql.md]: ooh. Not in this PR, but we need to document how those types get generated. Is that anywhere in the docs currently, @gjtorikian  ? (like have you found it? I don't think it is. I know what to write though)

File: docs/developer/putting-it-together.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
+Your ideal delivery blueprint spans delivery flow, generators, editors
+and other commands. All we need is something to pull it together.
+
+Your event listeners need to be invoked by Atomist handlers. The
+`SoftwareDeliveryMachine` takes care of this, ensuring that the
+correct handlers are emitted for use in `atomist.config.ts`, without
```
Comment (jessitron) [inline docs/developer/putting-it-together.md]: wait. We eliminated atomist.config.ts in favor of index.ts

File: docs/developer/putting-it-together.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+            DescribeStagingAndProd,
+            () => disposeProjectHandler,
+        )
+        .addSupportingEvents(OnDryRunBuildComplete);
+```
+
+The `SoftwareDeliveryMachine` instance will create the necessary
+Atomist event handlers to export.
+
+In `atomist.config.ts` you can bring them in simply as follows:
```
Comment (jessitron) [inline docs/developer/putting-it-together.md]: index.ts

File: docs/developer/respond-to-issue-creation.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+insufficient detail in the body, via a direct message in Slack, and
+provides them with a helpful link to the issue. Note that we make use
+of the person available via the `openedBy` field:
+
+```typescript
+export async function requestDescription(inv: NewIssueInvocation) {
+    if (!inv.issue.body || inv.issue.body.length < 10) {
+        await inv.context.messageClient.addressUsers(
+            `Please add a description for new issue ${inv.issue.number}: _${inv.issue.title}_: ${inv.id.url}/issues/${inv.issue.number}`,
+            inv.issue.openedBy.person.chatId.screenName);
```
Comment (jessitron) [inline docs/developer/respond-to-issue-creation.md]: not necessary for this PR, but this is not null safe. We don't have a chatId for every person

File: docs/developer/respond-to-issue-creation.md [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+}
+```
+
+This is registed with a `SoftwareDeliveryMachine` instance as follows:
+
+```typescript
+sdm.addNewIssueListeners(requestDescription)
+```
+
+Using the `credentials` on the `NewIssueInvocation`, you can easily
```
Comment (jessitron) [inline docs/developer/respond-to-issue-creation.md]: personal preference: can we not say "easily"
like ever (it makes people feel dumb sometimes when they don't already know how to do the thing)
including here, because there's more to using the GitHub API than having credentials. "you can use the GitHub API" is accurate.

File: docs/developer/respond-to-repo-creation.md [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        `A new repo was created: \`${i.id.owner}:${i.id.repo}\``, "general");
+};
+
+```
+
+Tagging a repo with topics based on its content is a useful
+action. `tagRepo` is a convenient function to construct a
+`ProjectListener` for this. It tags as an argument a `Tagger`, which
+looks at the project content and returns a `Tags` object. The
+following example from `atomist.config.ts` tags Spring Boot repos,
```
Comment (jessitron) [inline docs/developer/respond-to-repo-creation.md]: i think it goes in machine.ts?

File: docs/developer/set-goals.md (hunk 3 above)
Comment (jessitron) [reply docs/developer/set-goals.md]: Push Tests also apply to code transforms, and sometimes other things, so it makes sense for them to have their own section

File: docs/developer/set-goals.md [hunk 11]
```diff
@@ -79,11 +79,47 @@ The DSL includes support for logical operations. For example, this will build al
...
+
+#### Computed Values
+
+You can use computed `boolean` values or the results of synchronous or
+asynchronous functions returning `boolean` in the DSL, making it
+possible to bring in any state you wish. For example:
+
+```typescript
+whenPushSatisfies(IsMaven, HasSpringBootApplicationClass,
+	deploymentsToday < 25)
```
Comment (jessitron) [inline docs/developer/set-goals.md]: that's a constant boolean.
Do you mean `async (inv) => deploymentsToday < 25` ?

File: docs/developer/set-goals.md [hunk 12]
```diff
@@ -79,11 +79,47 @@ The DSL includes support for logical operations. For example, this will build al
...
+    .then(
+        given<Goals>(IsExpress).itMeans("express")
+            .compute(() => count++)
+            // Go into tree branch rule set
+            .then(
+                whenPushSatisfies(count > 0).itMeans("nope").setGoals(NoGoals),
+				whenPushSatisfies(TruePushTest).itMeans("yes").setGoals(HttpServiceGoals),
+            ),
+    );
+```
```
Comment (jessitron) [inline docs/developer/set-goals.md]: wow
Even I don't know what is going on there

Hey @johnsonr is this something we use and want to document?

File: docs/developer/transform.md (hunk 1 above)
Comment (jessitron) [reply docs/developer/transform.md]: It originally referred to this one I think: https://github.com/atomist/sdm/blob/0a8d8f488b83bfb6793f64c08870c33ea4c51699/lib/api-helper/command/transform/chattyDryRunAwareEditor.ts#L54

but @cdupuis may have info on a better way to do it now

File: docs/developer/transform.md [hunk 13]
```diff
@@ -15,6 +15,16 @@ This page shows how to:
...
+
+!!! tip
+    Dry run transforms are another example of how commands and events can work
+    hand in hand with Atomist to provide a uniquely powerful solution.
```
Comment (jessitron) [inline docs/developer/transform.md]: can we add, "The pull request is created in response to the transform command. It is merged in response to a successful build event.

**Comment by jessitron** [comment]
Thank you for gardening this!

**Comment by cdupuis** [comment]
I haven’t reviewed this yet but I think there are some questions around two different t concepts:

• build-aware transforms, are transforms that push to a branch, then wait for a build to pass before raising a PR on said branch for the changes to merge back into base. If the build fails, the build-aware transform can create an issue via the issueRouter. 
• dry-run transforms, are no special transform. It just means the transform is invoked with —dry-run=true to test the changes. The resulting changes aren’t committed but just printed as a git diff back to requester. 

I think this PR using dry-run transform where should actually refer to build-aware transforms.

File: docs/developer/putting-it-together.md (hunk 6 above)
Comment (gjtorikian) [reply docs/developer/putting-it-together.md]: I'm not sure; this text (as with most of the new content) was just taken [directly from what existed](https://github.com/atomist/docs/pull/286/files#diff-cacc5a8626a2cc71344704d473c67b79L442). I think you're probably right and this just hadn't been updated.

File: docs/developer/respond-to-issue-creation.md (hunk 8 above)
Comment (gjtorikian) [reply docs/developer/respond-to-issue-creation.md]: I'll update it just for the sake of not forgetting it later.

File: docs/developer/respond-to-issue-creation.md (hunk 9 above)
Comment (gjtorikian) [reply docs/developer/respond-to-issue-creation.md]: Totally agree. 👍 (This was also just a direct cut/copy of [what existed](https://github.com/atomist/docs/pull/286/files#diff-cacc5a8626a2cc71344704d473c67b79L268)).

File: docs/developer/respond-to-repo-creation.md (hunk 10 above)
Comment (gjtorikian) [reply docs/developer/respond-to-repo-creation.md]: 👍

File: docs/developer/transform.md (hunk 13 above)
Comment (gjtorikian) [reply docs/developer/transform.md]: 👍

**Comment by jessitron** [comment]
I vote we create an issue with cd's comment, and merge this PR, because that bit is present already
@gjtorikian

**Comment by gjtorikian** [comment]
@cdupuis: does that jive with you?

File: docs/developer/sdm-concepts.md [hunk 14]
```diff
@@ -1,133 +1,17 @@
...
+The heart of Atomist is its event handling. We're going to take a look
+at some of the core concepts and functionality
+that demonstrates the strengths of an SDM over traditional CI/CD pipelines.
 
 ## SDM process lifecycle
 
-The SDM lifecycle will be familiar to those developing
-persistent applications.
+The SDM lifecycle ought to be familiar to the flow of any persistent
+application:
```
Comment (jrday) [inline docs/developer/sdm-concepts.md]: The wording of this phrase trips me up. I _think_ we're trying to say that the SDM lifecycle should be familiar to anyone who has worked with persistent applications?

File: docs/developer/sdm-concepts.md [hunk 15]
```diff
@@ -191,260 +96,51 @@ sure the SDMs have different names.
...
+* Scanning code for security or quality issues on every push
+* Driving deployments and promotion between environments
+* Performing custom actions on deployment, such as kicking off
+    integration test suites.
 
-Using the `credentials` on the `NewIssueInvocation`, you can easily
-use the GitHub API to modify the issue, for example correcting
-spelling errors.
+The Atomist correlated event model also enables Atomist to provide you
+with visibility throughout the commit to deployment flow, in Slack or
```
Comment (jrday) [inline docs/developer/sdm-concepts.md]: "in Slack or Microsoft Teams"

**Comment by cdupuis** [comment]
> @cdupuis: does that jive with you?

yepp.
